### ISSUE TITLE
use_all_cores silently overrides concurrency

### ISSUE THREAD

**Issue by gcp** [glinscott/fishtest#1184]
If you launch the client once with ```use_all_cores```, it will save this in the configuration. If you then relaunch it with a lower ```--concurrency```, that will get ignored because the saved ```use_all_cores=True``` overrules it.

I can fix this, but I'm not sure what the desired behavior is. I think the easiest and least mistake prone is to warn and refuse to run (after all, these options directly conflict). But perhaps the intent of ```use_all_cores``` is to serve as an extra warning/unlock to not use concurrency=number of cpus, and simply shouldn't be saved to the config file at all, perhaps renaming it to ```allow_all_cores``` to reflect that.

Thoughts?

**Comment by ppigazzini** [comment]
The default behavior is to save 1 thread for OS operations or very light user activities. With systems fully dedicated to fishtest it's fine to use all available threads (actually noob asked for this flag to run his ChessDBCN fleet without patching the fishtest code). Perhaps the best options is to set `use_all_cores=False` when setting **only** `--concurrency`.
Changing that flag name could break some users setup (systemd unit file, launcher script etc.).

**Comment by vdbergh** [comment]
IMHO if an explicit concurrency is set then the logical behaviour is for it to take precedence over `--use_all_cores=True`.

**Comment by gcp** [comment]
The problem is that we recover some flags from the commandline and some from the config file, so ``` if an explicit concurrency is set``` becomes tricky, because we might restore it from the config file.

**Comment by ppigazzini** [comment]
`optparse` deprecated, replaced by `argparse` in python 3.2
https://docs.python.org/3/library/optparse.html
https://docs.python.org/3/library/argparse.html

**Comment by vdbergh** [comment]
`if an explicit concurrency is set` seems unambiguous to me. It might be either from the command line or from the config file with the proviso that command line parameters take precedence over the config file. 

Where IMHO is an issue is in Step 4 of parameter handling (conflict resolution). 
```
       if options.use_all_cores == "True":
            cpu_count = multiprocessing.cpu_count()
        else:
            cpu_count = int(options.concurrency)
            if cpu_count > multiprocessing.cpu_count() - 1:
                print(
                    (
                        "\nYou cannot have concurrency {} but at most:\n"
                        "{} with option --concurrency\n"
                        "{} with option --use_all_cores\n"
                    ).format(
                        options.concurrency,
                        multiprocessing.cpu_count() - 1,
                        multiprocessing.cpu_count(),
                    )
                )
                return None

```
The worker should set
```
cpu_count = int(options.concurrency)
```
and then bail out if this conflicts with the setting of `use_all_cores`.

**Comment by vdbergh** [comment]
IMHO the semantics of `use_all_cores` is: _allow_ the worker to use all cores and set the _default_ concurrency to the maximum number of cores.

**Comment by vdbergh** [comment]
Ok. I see. Then we create the opposite problem. If `--concurrency` is set in the config file then it will overrule `--use_all_cores=True` on the command line which would very confusing.

**Comment by gcp** [comment]
Maybe the simplest fix is to check if ```concurrency == cpu_count``` and if not, then unset ```use_all_cores```.

**Comment by vdbergh** [comment]
Maybe we should introduce `--allow_all_cores` in the config file (doing away with `--use_all_cores` ) and translate `--use_all_cores=True` on the command line into `--allow_all_cores=True --concurrency=<cpu_count>`.

**Comment by gcp** [comment]
That is backwards incompatible though, and leaves you without any easy way to send the same config out to a fleet of machines where you don't know the exact hardware in advance.

I don't see any downside to what I proposed (imply the ```use_all_cores=False```) so I'll make a pull request with that.

**Comment by gcp** [comment]
To make it clear: the reason why it works is that ```use_all_cores``` will always mean that we have written ```concurrency=<max>``` to the config file previously, so there's no concern that specifying ```use_all_cores``` next time will implicitly overwrite ```concurrency``` to the wrong value.

**Comment by gcp** [comment]
As @ppigazzini pointed out, this doesn't work because a stored concurrency < max would also override the command line.

Maybe changing the logic so it knows the difference between command line options and options from the config file is preferable (for any solution that wants to keep backwards compat).

**Comment by vdbergh** [comment]
This is fixed by #1189. So can be closed!
